### PR TITLE
Added the configuration file .clang-format to use with clang-format

### PR DESCRIPTION
**Pull request by nnormand**
# PR Description

This PR adds the configuration file `.clang-format` for `clang-format`. The standalone tool `clang-format` can be used to format C++. It is documented [here](http://clang.llvm.org/docs/ClangFormat.html#standalone-tool) and can be installed from the clang distribution ([clang homepage](http://clang.llvm.org)). Once installed, `clang-format` is fairly easy to use:
`clang-format file` (output on the standard output) or `clang-format -i file` (inline formatting of `file`) with `.clang-file` in the directory or a a parent directory of the input file.

The set of rules were derived from DGtal [coding style](https://github.com/DGtal-team/DGtal/blob/master/CONTRIBUTING.md#coding-style) and from the DGtal source code base. `clang-rules` rules are available [here](http://clang.llvm.org/docs/ClangFormatStyleOptions.html#configurable-format-style-options). Also note that `clang-format` doesn't address all DGtal coding style instructions (_e.g._ variable naming) but can help creating a reasonably formatted code base.

# Checklist

- [x] Unit-test of your feature with [Catch](http://dgtal.org/doc/stable/moduleCatch.html).
- [x] Doxygen documentation of the code completed (classes, methods, types, members...)
- [x] Documentation module page added or updated.
- [x] New entry in the [ChangeLog.md](https://github.com/DGtal-team/DGtal/blob/master/ChangeLog.md) added.
- [x] No warning raised in Debug ```cmake``` mode (otherwise, Travis C.I. will fail).
- [x] All continuous integration tests pass (Travis & appveyor)

### CODE REVIEW

File: .clang-format [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+# - fonction parameters on the open bracket,
+# - trailing comments.
+# Don't align:
+# - consecutive assignments or declarations,
+# - operands of binary and ternary expressions.
+#
+AlignAfterOpenBracket: Align
+AlignConsecutiveAssignments: false
+AlignConsecutiveDeclarations: false
+AlignOperands: false
```
Comment (dcoeurjo) [inline .clang-format]: I do like this one;)

Comment (dcoeurjo) [inline .clang-format]: (when set to true)

**Comment by dcoeurjo** [comment]
Thanks for the PR, I'm on it with @kerautret

File: .clang-format [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+AllowShortFunctionsOnASingleLine: false
+AllowShortIfStatementsOnASingleLine: false
+AllowShortLoopsOnASingleLine: false
+#
+# Line breaks
+# -----------
+#
+# Insert line breaks:
+# - before braces (Allman style),
+# - after the function definition return type,
```
Comment (dcoeurjo) [inline .clang-format]: I prefer to have the return type in line if we have space (clang Penalty.. stuff for the 80col autofill)

**Comment by dcoeurjo** [comment]
I've slightly modified the format (cf my test file https://gist.github.com/dcoeurjo/918f84fe1151accf8fd6a9ae1af9afb8)

What do you think ?

Two issues I still see:

* the doxygen comment is not ok (missing ' ') when clamped to 80col
* it would be nice to have "<<" aligned but I don't know if it is possible.

**Comment by dcoeurjo** [comment]
@nnormand would you have an idea for the two issues I have ? Are my edits ok for you?

**Comment by nnormand** [comment]
hi @dcoeurjo

I fully agree with your changes, the code looks nicer. However, for some reason, formatting with these new rules tend to induce more changes than before.

~~~
git checkout 85a146aa249b592c8d4c4d1eabd4e1b79405ca4e
(c=0;for f in $(find . -name '*.h' -o -name '*.ih' -o -name '*.cpp');do d=$(leven "$(expand -4 $f|perl -pe 's/\h+$//')" "$(clang-format $f)" || echo 0);echo "$f: $d";c=$(($c+$d));done;echo "total: $c")|tee 85a146aa249b592c8d4c4d1eabd4e1b79405ca4e.clang-format.log
git checkout 10c06d2d9fba327c8a248c7e9efa6130756aaa4e
(c=0;for f in $(find . -name '*.h' -o -name '*.ih' -o -name '*.cpp');do d=$(leven "$(expand -4 $f|perl -pe 's/\h+$//')" "$(clang-format $f)" || echo 0);echo "$f: $d";c=$(($c+$d));done;echo "total: $c")|tee 10c06d2d9fba327c8a248c7e9efa6130756aaa4e.clang-format.log
~~~

I'm not sure I understand your first issue and I don't know if the second one can be addressed.

**Comment by dcoeurjo** [comment]
For the changes w.r.t. to the existing code, I'm not so really worried about that. Most files do not comply with the CodingStyle or formatting rules. 

Let's create "the format we want" in this PR and we'll see later if we decide to apply it to the existing code (or just ask the contributors to comply for their new PRs).

**Comment by dcoeurjo** [comment]
(nice pre-commit script: https://dx13.co.uk/articles/2015/4/3/Setting-up-git-clang-format.html)

**Comment by dcoeurjo** [comment]
@JacquesOlivierLachaud , @kerautret  can you please have a look to this clang-format ?

**Comment by kerautret** [comment]
@dcoeurjo I try but I have this error : (I miss something  ?)

YAML:14:24: error: invalid boolean
AlignAfterOpenBracket: Align
                       ^~~~~
Error reading /Users/kerautre/.clang-format: Invalid argument
Can't find usable .clang-format, using LLVM style

**Comment by kerautret** [comment]
that looks fine expect for the boost program_options but we could deactivate it with locally with ```// clang-format off```

**Comment by dcoeurjo** [comment]
ok thanks. Merging this one.

**Pull request merged**